### PR TITLE
Adding in FSDP Docs

### PR DESCRIPTION
**Pull request by bcui19**
# What does this PR do?

Adding FSDP docs to composer

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: docs/source/notes/distributed_training.rst [hunk 1]
```diff
@@ -135,5 +135,91 @@ from the trainer.
...
+
+The full spec and defaults for Composer's `fsdp_config` is here:
+
+.. code:: python
+
+    fsdp_config = {
+      'sharding_strategy': str = 'FULL_SHARD' | 'SHARD_GRAD_OP' | 'NO_SHARD', # Default: 'FULL_SHARD'
+      'min_params': float # Default: 1e8
+      'cpu_offload': bool = True | False, # Default: False, cpu_offload not supported yet
+      'mixed_precision': str = 'full' | 'default' | 'pure', # Default: 'default'
```
Comment (abhi-mosaic) [inline docs/source/notes/distributed_training.rst:158]:
```suggestion
      'mixed_precision': str = 'FULL' | 'DEFAULT' | 'PURE', # Default: 'DEFAULT'
```

**Comment by abhi-mosaic** [comment]
@bcui19 here's a pseudocode snippet that can be used as an example for FSDP:

```python
import torch
from composer import Trainer

class Block (nn.Module):
    ...

class Model(nn.Module):
    def __init__(self, n_layers):
        super().__init__()
        self.blocks = nn.ModuleList([
            Block(...) for _ in range(n_layers)
        ]),
        self.head = torch.nn.Linear(...)

    def forward(self, inputs):
        ...

    # FSDP Wrap Function
    def fsdp_wrap_fn(self, module):
        return isinstance(module, Block)

    # Activation Checkpointing Function
    def activation_checkpointing_fn(self, module):
        return isinstance(module, Block)


class MyComposerModel(ComposerModel):

    def __init__(self, n_layers):
        super().__init__()
        self.model = Model(n_layers)
        ...

    def forward(self, batch):
        ...

    def eval_forward(self, batch, outputs=None):
        ...

    def loss(self, outputs, batch):
        ...

    ...


composer_model = MyComposerModel(n_layers=3)

fsdp_config = {
    'sharding_strategy': 'FULL_SHARD',
    'min_params': 1e8,
    'cpu_offload': False, # Not supported yet
    'mixed_precision': 'DEFAULT',
    'backward_prefetch': 'BACKWARD_POST',
    'activation_checkpointing': False,
    'activation_cpu_offload': False,
    'verbose': True
}


trainer = Trainer(
    model=composer_model,
    fsdp_config=fsdp_config,
    ...
)

trainer.fit()
```

File: docs/source/notes/distributed_training.rst [hunk 2]
```diff
@@ -18,9 +18,9 @@ performing the same work, so inspecting the rank zero is sufficient to
 reason about memory, performance, and other properties.
 
 Within Composer, we have two options for data-parallelism-only
-execution: `Pytorch DDP`_` and `DeepSpeed Zero`_. We currently default to
+execution: `Pytorch DDP`_, `DeepSpeed Zero`_, and `Pytorch FSDP`_. We currently default to
 Pytorch DDP, though DeepSpeed Zero can provide better performance and
-lower memory utilization when configured correctly.
+lower memory utilization when configured correctly. # TODO: add some commentary on FSDP
```
Comment (abhi-mosaic) [inline docs/source/notes/distributed_training.rst:23]: Can we reorder these and emphasize FSDP?

File: docs/source/notes/distributed_training.rst [hunk 3]
```diff
@@ -135,5 +135,91 @@ from the trainer.
...
+      # Note: you can explictly provide a dictionary too
+      # 'mixed_precision': dict = {
+      #   'param_dtype': 'fp32' | 'fp16' | 'bf16',
+      #   'reduce_dtype': 'fp32' | 'fp16' | 'bf16',
+      #   'buffer_dtype': 'fp32' | 'fp16' | 'bf16',
+      # },
+      'backward_prefetch': str = 'BACKWARD_PRE' | 'BACKWARD_POST' | 'NONE', # Default: 'BACKWARD_POST'
+      'activation_checkpointing': bool = True | False, # Default: False
+      'activation_cpu_offload': bool = True | False, # Default: False
+    }
```
Comment (abhi-mosaic) [inline docs/source/notes/distributed_training.rst:168]: Need to add `verbose: bool = True | False,` that's on me I forgot about it previously.

**Review by abhi-mosaic — APPROVED** [review]
LGTM, thanks @bcui19 !

File: docs/source/notes/distributed_training.rst [hunk 4]
```diff
@@ -135,5 +136,160 @@ from the trainer.
...
+2) If the root module (e.g. `GPT`) defines a function :code:`def fsdp_wrap_fn(module: torch.nn.Module) -> bool`, then that function will be used to evaluate the root module's children.
+3) If any module has more parameters than :code:`fsdp_config['min_params']`, it will be wrapped.
+
+These rules are meant to make it easy for users to modify existing models for usage with FSDP. You can either add attributes to modules you want to wrap (#1), define a filter (#2), or make no changes at all and just use the size-based policy via :code:`fsdp_config['min_params'] = ...` (#3).
+
+In `gpt.py <https://github.com/mosaicml/benchmarks/blob/6972fe3000d5a5480d8757ff710965514155e8db/llm/llm/gpt.py>`__, you can see that `we used rule #2 <https://github.com/mosaicml/benchmarks/blob/6972fe3000d5a5480d8757ff710965514155e8db/llm/llm/gpt.py#L172>` to specify that all :code:`GPTBlock` modules within :code:`GPT` should be wrapped. Alternatively, we could have easily attributed each of the blocks with :code:`block._fsdp_wrap = True` and it would have accomplished the same thing. Whatever style you prefer, it's up to you!
```
Comment (bandish-shah) [inline docs/source/notes/distributed_training.rst:285]: @bcui19 the link for "we used rule #2" is broken here, missing a "__" after the URL

**Pull request merged**